[processor/transform] introduce aggregate_on_attributes function for metrics (#33334)
open-telemetry/opentelemetry-collector-contrib: processor/transformprocessor/README.md

- `copy_metric(desc="new desc") where description == "old desc"`

### aggregate_on_attributes

`aggregate_on_attributes(function, Optional[attributeSet])`

The `aggregate_on_attributes` function aggreates all metrics having the attributes present in `attributeSet` using the aggregation function specified in `function`. Parameter `attributeSet` is optional and if it's not present, all attributes of the metric are taken into consideration for the aggregation function.

The function supports the following data types:

- sum
- gauge
- histogram
- exponential histogram

Supported aggregation functions are:

- sum
- max
- min
- mean
- median

> **NOTE:** Only the `sum` aggregation function is supported for histogram and exponential histogram datatypes.

Author: Agree, but `keep_matching_keys` is not merged yet, should we include it? At least in this case I cannot reference the function yet.

- `copy_metric(desc="new desc") where description == "old desc"`

### aggregate_label

`aggregate_label(function, labelSet)`

The `aggregate_label` function aggreates all metrics having the labels present in `labeSet` using the aggregation function specified in `function`.

The function supports the following data types:

- sum
- gauge
- histogram

> - exponantial histogram

Reviewer: ```suggestion
- exponential histogram
```

Supported aggregation functions are:

- sum
- max
- min
- mean
- median

**NOTE:** Only the `sum` agregation function is supported for histogram and exponential histogram datatypes.

Examples:

- `aggregate_label(sum, [label1, label2]) where name == "system.memory.usage`

### aggregate_label_value

`aggregate_label_value(function, label, labelSet, newValue)`

> The `aggregate_label_value` function aggreates all metrics having the label present in `labeSet` with one of the values present in `valueSet` using the aggregation function specified in `function`. Additionally it substitues the values of `label` present in `valueSet` with a new value specified by `newValue`.

Reviewer: ```suggestion
The `aggregate_label_value` function aggreates all metrics having the label present in `labeSet` with one of the values present in `valueSet` using the aggregation function specified in `function`. Additionally it substitutes the values of `label` present in `valueSet` with a new value specified by `newValue`.
```

The function supports the following data types:

- sum
- gauge
- histogram

> - exponantial histogram

Reviewer: ```suggestion
- exponential histogram
```

- `copy_metric(desc="new desc") where description == "old desc"`

> ### aggregate_on_attributes

Reviewer 2: I think the docs need to explain a little more about what it means to aggregate a metric across attributes.  Or link to a doc that does explain the concept.

Author: adapted, please check if it's enough

> `aggregate_on_attributes(function, Optional[attributeSet])`

Reviewer: ```suggestion
`aggregate_on_attributes(function, Optional[attributes])`
```
What do you think about removing `set` from the name? To me, `set` implies it may be a map.

> The `aggregate_on_attributes` function aggreates all metrics having the attributes present in `attributeSet` using the aggregation function specified in `function`. Parameter `attributeSet` is optional and if it's not present, all attributes of the metric are taken into consideration for the aggregation function.

Reviewer: ```suggestion
The `aggregate_on_attributes` function aggregates all metrics having the attributes present in `attributeSet` using the aggregation function specified in `function`. `function` is a string with a supported aggregation function, and `attributeSet` is an optional list of attributes to use for aggregation. If `attributeSet` is not specified, all attributes of the metric are taken into consideration for aggregation.
```

> The function supports the following data types:

Reviewer: ```suggestion
The following metric types can be aggregated:
```

> The `aggregate_on_attributes` function aggregates all metrics having the attributes present in `attributeSet` using the aggregation function specified in `function`. `function` is a string with a supported aggregation function, and `attributeSet` is an optional list of attributes to use for aggregation. If `attributeSet` is not specified, all attributes of the metric are taken into consideration for aggregation.

Author: ```suggestion
The `aggregate_on_attributes` function aggregates all metrics having the attributes present in `attributeSet` using the aggregation function specified in `function`. `function` is a string with a supported aggregation function, and `attributes` is an optional list of attributes to use for aggregation. If `attributes` is not specified, all attributes of the metric are taken into consideration for aggregation.
```

> The `aggregate_on_attributes` function aggregates all metrics having the attributes present in `attributes` using the aggregation function specified in `function`. `function` is a string with the name of a supported aggregation function, and `attributes` is an optional list of attributes to use for aggregation. If `attributes` is not specified, all attributes of the metric are taken into consideration for aggregation.

Reviewer 2: ```suggestion
The `aggregate_on_attributes` function aggregates all datapoints in the metric based on the supplied attributes. `function` is a string that represents the function to use for the aggregation and `attributes` is an optional list of attribute keys to aggregate upon. If `attributes` is not specified, all attributes of the metric are taken into consideration for aggregation.
```

- sum
- gauge
- histogram
- exponential histogram

Supported aggregation functions are:

- sum
- max
- min
- mean
- median
- count

**NOTE:** Only the `sum` aggregation function is supported for histogram and exponential histogram datatypes.

Examples:

- `aggregate_on_attributes("sum", [attr1, attr2]) where name == "system.memory.usage`
- `aggregate_on_attributes("max") where name == "system.memory.usage`

The `aggregate_on_attributes` function can also be used in conjunction with 
[keep_matching_keys](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs#keep_matching_keys) or
[delete_matching_keys](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs#delete_matching_keys).

> To leave out all attribute keys matching the desired regex and aggregate the metrics afterwards, you can perform the following statement sequence:

Reviewer: ```suggestion
For example, to remove attribute keys matching a regex and aggregate the metrics on the remaining attributes, you can perform the following statement sequence:
```

```yaml
statements:
   - delete_matching_keys(attributes, "(?i).*myRegex.*") where name == "system.memory.usage
   - aggregate_on_attributes("sum") where name == "system.memory.usage
```

> The same approach is possible also for the function keep_matching_keys.

Reviewer: ```suggestion
To aggregate only using a specified set of attributes, you can use `keep_matching_keys`.
```

- `scale_metric(0.1)`: Scale the metric by a factor of `0.1`. The unit of the metric will not be modified.
- `scale_metric(10.0, "kWh")`: Scale the metric by a factor of `10.0` and sets the unit to `kWh`.

### aggregate_on_attributes

`aggregate_on_attributes(function, Optional[attributes])`

> The `aggregate_on_attributes` function aggregates all datapoints in the metric based on the supplied attributes. To have the datapoints aggregated, each datapoint has to contain all of the specified attribute keys. If one of the key is missing in the attributes list, the datapoint will be left out from the aggregation. If no attributes are specified, all attributes of the metric datapoints are taken into consideration for aggregation, which means that every datapoint needs to have the same keys present in its attribute list.

Reviewer 2: Something doesn't seem right about this paragraph. It seems to be suggesting that datapoints that have attributes don't have matching attributes will be left alone, but those datapoints will have any non-matching attributes removed and they'll be aggregated together right?

My understanding of the function was that it dropped all attributes that were not in the supplied attributes list and then aggregated the datapoints based on any remaining attributes.  That means that if `attributes` is empty, every attribute is dropped and all datapoints are combined into one.

> `function` is a string that represents the function to use for the aggregation and `attributes` is an optional list of attribute keys to aggregate upon.

Reviewer: ```suggestion
`function` is a case-sensitive string that represents the function to use for the aggregation and `attributes` is an optional list of attribute keys to aggregate upon.
```
